include/revng/RestructureCFG/ScopeGraphGraphTraits.h

```diff
@@ -0,0 +1,232 @@
+      llvm::GraphTraits<Scope<llvm::BasicBlock *>>::verify(&N);
+    }
+  }
+};
+
+template<>
+struct llvm::GraphTraits<Scope<const llvm::Function *>>
+  : public llvm::GraphTraits<Scope<const typename llvm::BasicBlock *>> {
+  using NodeRef = const llvm::BasicBlock *;
+  using nodes_iterator = pointer_iterator<Function::const_iterator>;
+
+  static NodeRef getEntryNode(Scope<const llvm::Function *> G) {
+    return &G.Graph->getEntryBlock();
+  }
+
+  static nodes_iterator nodes_begin(Scope<const llvm::Function *> G) {
+    return nodes_iterator(G.Graph->begin());
+  }
+
+  static nodes_iterator nodes_end(Scope<const llvm::Function *> G) {
+    return nodes_iterator(G.Graph->end());
+  }
+
+  static size_t size(Scope<const llvm::Function *> G) {
+    return G.Graph->size();
+  }
+
+  // Add a verify method to the trait which invokes the `verify` of the
+  // `BasicBlock *` trait for each node in the graph
+  static void verify(Scope<const llvm::Function *> G) {
+    for (auto &N : *G.Graph) {
+      llvm::GraphTraits<Scope<const llvm::BasicBlock *>>::verify(&N);
+    }
+  }
+};
+
+inline std::string getNodeLabel(const llvm::BasicBlock *N,
+                                const llvm::Function *F) {
+  if (N->hasName()) {
+    return N->getName().str();
+  } else {
+
+    // We iterate over the function and we assign an incremental ID for each
+    // unnamed `BasicBlock` that we encounter
+    unsigned ID = 0;
+    for (const auto *Node : llvm::nodes(F)) {
+      if (N == Node) {
+        return "block_" + std::to_string(ID);
+      } else {
+        ++ID;
+      }
+    }
+  }
+
+  revng_abort("We should always find the basicblock in the loop above");
+}
+
+template<>
+struct llvm::DOTGraphTraits<Scope<llvm::Function *>>
+  : public llvm::DefaultDOTGraphTraits {
+  using llvm::DefaultDOTGraphTraits::DefaultDOTGraphTraits;
+
+  std::string getNodeLabel(const llvm::BasicBlock *N,
+                           const Scope<llvm::Function *> G) {
+    return ::getNodeLabel(N, G.Graph);
+  }
+
+  // TODO: we may want to specialize `getEdgeAttributes` method in order to
+  //       print (also in the `const` version below):
+  //       - The scope closer edges should be markerd with a different color
+  //       - The CFG edges that are ignored due to the fact that they are
+  //         cancelled by a goto matker should be dashed, and should not afftect
+  //         the layout of the graph [constraint=false]
+};
+
+template<>
+struct llvm::DOTGraphTraits<Scope<const llvm::Function *>>
+  : public llvm::DefaultDOTGraphTraits {
+  using llvm::DefaultDOTGraphTraits::DefaultDOTGraphTraits;
+
+  std::string getNodeLabel(const llvm::BasicBlock *N,
+                           const Scope<const llvm::Function *> G) {
+    return ::getNodeLabel(N, G.Graph);
+  }
+};
+
+// Debug function used dump a serialized representation of the `ScopeGraph` on a
+// stream
+template<typename O>
+debug_function void dumpScopeGraph(O &Output, llvm::Function &F) {
```

Comment:
As we agreed, I dropped the `Output` argument https://github.com/revng/revng/blob/fe8a4d7ebc61d0ceb8d64cdfe5c3ecf8fd1c21f4/include/revng/RestructureCFG/ScopeGraphGraphTraits.h#L164 and just used `llvm::dbgs()`, and dropped the logger from the `ScopeGraphLoggerPass` pass.

In addition, this is the only "non graph related" function in this header, but I cannot move it into `ScopeGraphUtils` because this needs the `ScopeGraph` `GraphTraits` to iterate over nodes and children, and that would induce a circular dependency. Let me know if you want a separate header and relative `.cpp` just for this or if it is ok to leave this in this header